docs/developers/tutorials/staking-contract.md

```diff
@@ -0,0 +1,1416 @@
+
+For this reason, we have to add some additional checks. The endpoint implementations will have to be changed to the following (the rest of the code remains the same):
+```rust
+    #[payable("EGLD")]
+    #[endpoint]
+    fn stake(&self) {
+        let payment_amount = self.call_value().egld_value();
+        require!(payment_amount > 0, "Must pay more than 0");
+
+        let caller = self.blockchain().get_caller();
+        let stake_mapper = self.staking_position(&caller);
+
+        let new_user = self.staked_addresses().insert(caller.clone());
+        let mut staking_pos = if !new_user {
+            stake_mapper.get()
+        } else {
+            let current_block = self.blockchain().get_block_epoch();
+            StakingPosition {
+                stake_amount: BigUint::zero(),
+                last_action_block: current_block,
+            }
+        };
+
+        self.claim_rewards_for_user(&caller, &mut staking_pos);
+        staking_pos.stake_amount += payment_amount;
+
+        stake_mapper.set(&staking_pos);
+    }
+
+    #[endpoint]
+    fn unstake(&self, opt_unstake_amount: OptionalValue<BigUint>) {
+        let caller = self.blockchain().get_caller();
+        self.require_user_staked(&caller);
+
+        let stake_mapper = self.staking_position(&caller);
+        let mut staking_pos = stake_mapper.get();
+
+        let unstake_amount = match opt_unstake_amount {
+            OptionalValue::Some(amt) => amt,
+            OptionalValue::None => staking_pos.stake_amount.clone(),
+        };
+        require!(
+            unstake_amount > 0 && unstake_amount <= staking_pos.stake_amount,
+            "Invalid unstake amount"
+        );
+
+        self.claim_rewards_for_user(&caller, &mut staking_pos);
+        staking_pos.stake_amount -= &unstake_amount;
+
+        if staking_pos.stake_amount > 0 {
+            stake_mapper.set(&staking_pos);
+        } else {
+            stake_mapper.clear();
+            self.staked_addresses().swap_remove(&caller);
+        }
+
+        self.send().direct_egld(&caller, &unstake_amount);
+    }
+
+    #[endpoint(claimRewards)]
+    fn claim_rewards(&self) {
+        let caller = self.blockchain().get_caller();
+        self.require_user_staked(&caller);
+
+        let stake_mapper = self.staking_position(&caller);
+        let mut staking_pos = stake_mapper.get();
+        self.claim_rewards_for_user(&caller, &mut staking_pos);
+
+        stake_mapper.set(&staking_pos);
+    }
+
+    fn require_user_staked(&self, user: &ManagedAddress) {
+        require!(self.staked_addresses().contains(user), "Must stake first");
+    }
+```
+
+For the `stake` endpoint, we provide a default entry is user was not previously staked. The `insert` method of `UnorderedSetMapper` returns `true` if the entry is new, `false` if the user was already in the list, so we can use that result instead of checking for `stake_mapper.is_empty()`.
```

Comment:
```suggestion
For the `stake` endpoint, in case the user was not previously staked, we provide a default entry. The `insert` method of `UnorderedSetMapper` returns `true` if the entry is new, `false` if the user was already in the list, so we can use that result instead of checking for `stake_mapper.is_empty()`.
```